pkg/controller/azvolume.go

```diff
@@ -81,260 +82,303 @@ func (r *ReconcileAzVolume) Reconcile(ctx context.Context, request reconcile.Req
+			if azVolume.Status.Detail.ResponseObject == nil {
+				azVolume.Status.Detail.ResponseObject = &v1alpha1.AzVolumeStatusParams{}
 			}
-			updated.Status.Detail.ResponseObject.CapacityBytes = status.CapacityBytes
-			updated.Status.Detail.ResponseObject.NodeExpansionRequired = status.NodeExpansionRequired
+			azVolume.Status.Detail.ResponseObject.CapacityBytes = status.CapacityBytes
+			azVolume.Status.Detail.ResponseObject.NodeExpansionRequired = status.NodeExpansionRequired
 		} else {
-			updated.Status.Detail.ResponseObject = nil
+			azVolume.Status.Detail.ResponseObject = nil
 		}
 	}
+	return azVolume
+}
 
-	updated.Status.Detail.Phase = phase
-
-	if err := r.client.Update(ctx, updated, &client.UpdateOptions{}); err != nil {
-		klog.Errorf("failed to update status of AzVolume (%s): %v", volumeName, err)
-		return nil, err
+func (r *ReconcileAzVolume) updateStatusPhase(ctx context.Context, azVolume *v1alpha1.AzVolume, phase v1alpha1.AzVolumePhase) *v1alpha1.AzVolume {
+	if azVolume == nil {
+		return nil
 	}
-
-	return updated, nil
+	if azVolume.Status.Detail == nil {
+		azVolume.Status.Detail = &v1alpha1.AzVolumeStatusDetail{}
+	}
+	azVolume.Status.Detail.Phase = phase
+	return azVolume
 }
 
-func (r *ReconcileAzVolume) updateStatusWithError(ctx context.Context, volumeName string, azVolume *v1alpha1.AzVolume, err error) (*v1alpha1.AzVolume, error) {
+func (r *ReconcileAzVolume) updateError(ctx context.Context, azVolume *v1alpha1.AzVolume, err error) *v1alpha1.AzVolume {
 	if azVolume == nil {
-		var derr error
-		azVolume, derr = azureutils.GetAzVolume(ctx, r.client, r.azVolumeClient, volumeName, r.namespace, true)
-		if derr != nil {
-			klog.Errorf("failed to get AzVolume (%s): %v", volumeName, derr)
-			return nil, derr
-		}
+		return nil
+	}
+	azVolumeError := &v1alpha1.AzError{
+		ErrorCode:    util.GetStringValueForErrorCode(status.Code(err)),
+		ErrorMessage: err.Error(),
 	}
+	azVolume.Status.Error = azVolumeError
+	return azVolume
+}
 
-	updated := azVolume.DeepCopy()
-	if err != nil {
-		azVolumeError := &v1alpha1.AzError{
-			ErrorCode:    util.GetStringValueForErrorCode(status.Code(err)),
-			ErrorMessage: err.Error(),
+func (r *ReconcileAzVolume) updateWithRetry(ctx context.Context, azVolumeName string, updateFunc func(*v1alpha1.AzVolume)) error {
+	conditionFunc := func() (bool, error) {
+		azVolume, err := azureutils.GetAzVolume(ctx, r.client, r.azVolumeClient, azVolumeName, r.namespace, true)
+		if err != nil {
+			klog.Errorf("failed to get AzVolume (%s): %v", azVolumeName, err)
+			// if failed to get AzVolume then escape the wait
+			return false, err
 		}
-		updated.Status.Error = azVolumeError
 
-		if err := r.client.Update(ctx, updated, &client.UpdateOptions{}); err != nil {
-			klog.Errorf("failed to update error status of AzVolume (%s): %v", volumeName, err)
-			return nil, err
+		azVolume = azVolume.DeepCopy()
+		updateFunc(azVolume)
+
+		if err := r.update(ctx, azVolume); err != nil {
```

Comment:
Shouldn't we only retry on conflict? If `azVolume` is `nil` then `update` will always return `InvalidArguement` and we'll loop until timeout even though the error is not recoverable.